### DISCUSSION TITLE
customizable locale dictionary

### DISCUSSION THREAD

**Discussion by ariaieboy** [Laravel-Lang/common#13] [Q&A]
hello everyone
I had a problem with the Persian translation of this project, which was inconsistency in the translations.

For example, some keywords like `password` have multiple translations in Persian and each person can translate it to a specific word in Persian.

I plan to create a dictionary for Persian translation. so anyone who wants to help translate the keys has a guideline about selecting the correct translations for most used keywords so we have normalized translations across all of our Persian keys.

A good feature can be a customizable dictionary for these keywords so anyone could change the most used keywords based on the business guideline or the culture of their application.

**Comment by andrey-helldar** [comment]
I think that adding a dictionary of dialects is inappropriate because it will be difficult to maintain. Instead we can create new localizations like `fa_Pes` and `fa_Prs`. This is exactly what we have done for the German, Portuguese, Serbian, Uzbek and Chinese languages:

- [de](https://github.com/Laravel-Lang/lang/tree/main/locales/de)
- [de_CH](https://github.com/Laravel-Lang/lang/tree/main/locales/de_CH)
- [pt](https://github.com/Laravel-Lang/lang/tree/main/locales/pt)
- [pt_BR](https://github.com/Laravel-Lang/lang/tree/main/locales/pt_BR)
- [sr_Cyrl](https://github.com/Laravel-Lang/lang/tree/main/locales/sr_Cyrl)
- [sr_Latn](https://github.com/Laravel-Lang/lang/tree/main/locales/sr_Latn)
- [sr_Latn_ME](https://github.com/Laravel-Lang/lang/tree/main/locales/sr_Latn_ME)
- [uz_Cyrl](https://github.com/Laravel-Lang/lang/tree/main/locales/uz_Cyrl)
- [uz_Latn](https://github.com/Laravel-Lang/lang/tree/main/locales/uz_Latn)
- [zh_CN](https://github.com/Laravel-Lang/lang/tree/main/locales/zh_CN)
- [zh_HK](https://github.com/Laravel-Lang/lang/tree/main/locales/zh_HK)
- [zh_TW](https://github.com/Laravel-Lang/lang/tree/main/locales/zh_TW)

Thus, all the necessary options can be installed in the application.

And if in the final application you need to explicitly use only one localization, for example, `fa`, saving the necessary dialect in it, then you can use the alias setting in the installer:

```php
'aliases' => [
    'fa_Pes' => 'fa',
],
```
https://github.com/Laravel-Lang/publisher/blob/main/config/public.php#L50-L54

And the project will have three available localizations:

- fa
- fa_Pes
- fa_Prs

https://en.wikipedia.org/wiki/ISO_639_macrolanguage#fas

What do you think?

**Reply by ariaieboy** [reply]
actually what I have in mind is different than dialects.

maybe what I have in mind is more common in Persian than in any other language. 

It's about the synonym of words. not dialects. for example, `password` has at least 4 synonyms and we used 2 of the synonyms in laravel-lang. for the Persian language. 

first of all, I want to only see 1 synonym for each hot keyword in the whole project.

and second, if it's possible to have the option to switch one with another. 

for example, let's say we want only use `E-mail` in the whole project and we have the feature to change it to `email` if we want for the whole project.

**Reply by andrey-helldar** [reply]
Understood. It's more of a slang used in different parts of the country. For example, in the south of Germany, the expressions `Schnecken Morgen` and `Schneckseln Morgen` have the same meaning - `See you in the morning`. But in the northern part, `Schnecken Morgen` has the same meaning, but `Schneckseln Morgen` is translated as `let's fuck in the morning`.

I'm afraid that such changes will be extremely difficult to maintain, since at any time someone can make adjustments to laravel-lang, correcting the meaning of the code, and developers will change the words. In addition, for each specific language, you will have to choose your own slang.

BUT if you give the opportunity to translate entire phrases, then this problem can be solved. In practice, this will look like a separate file, like so:

```
php/
  - main.json
  - inline.json
  - slang/
    - foo/
      - main.json
      - inline.json
    - bar/
      - main.json
      - inline.json
json/
  - main.json
  - inline.json
  - slang/
    - foo/
      - main.json
      - inline.json
    - bar/
      - main.json
      - inline.json
```

In this case, we can create any number of slang dictionaries and check their existence during installation. If there is no translation value, take fallback, that is, from the main file.

**Reply by ariaieboy** [reply]
for now, if I found some free time in my schedule I am gonna start normalizing the translation for Persian and after that, I will look into this again

**Reply by andrey-helldar** [reply]
In this case, in the near future I will implement this structure in the update scripts, as well as the publisher.

The only thing I'm not sure is that the word "slang" should be used. In my opinion, "jargon" is better suited. But even here I'm not sure. Do you have any thoughts on this?

**Reply by andrey-helldar** [reply]
I figured it out. The structure does not need to be changed.

Current structure:

```
locales/
  af/
    _excludes.json
    json-inline.json
    json.json
    php-inline.json
    php.json
```

Next structure:

```
locales/
  af/
    _excludes.json
    json-inline.json
    json.json
    php-inline.json
    php.json
    variations/
      foo/
        json-inline.json
        json.json
        php-inline.json
        php.json
```

**Reply by andrey-helldar** [reply]
I will be able to add support for this functionality to the publisher only in January. Before that, I need to finish another project.

**Reply by andrey-helldar** [reply]
@ariaieboy, hey! I thought about various possible options for improving the logic and came to the conclusion that the best solution would be to create a new localization.

This is due to the fact that the publisher does not know anything about the packages connected to it. Thus, the publisher will not know about the available localization options not only in the Lang project, but also in any others that our team has nothing to do with (we even created a template for this).

Thus, each developer will have to manually go into the files of the project he is installing, see if there are options in it, and only after that configure the configuration file. It's complicated.

Therefore, I suggest creating new localizations with a variance mark. It will be right.
